### PR TITLE
Cherry pick bug fixes into 2.1

### PR DESCRIPTION
**Pull request by johnou**
Node recovery bug fixes from 2.0 60e1661..2f30188

### CODE REVIEW

File: protocols/raft/src/main/java/io/atomix/protocols/raft/service/RaftServiceContext.java [hunk 1]
```diff
@@ -418,11 +423,8 @@ private void maybeInstallSnapshot(long index) {
     serviceExecutor.execute(() -> {
       log.debug("Opening session {}", session.sessionId());
 
-      // Update the state machine index/timestamp.
-      tick(index, timestamp);
-
-      // If a snapshot exists prior to the given index and hasn't yet been installed, install the snapshot.
-      maybeInstallSnapshot(index);
+      // Update the session's timestamp to prevent it from being expired.
```
Comment (johnou) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/service/RaftServiceContext.java]: Corrected possible bad merge.

**Comment by johnou** [comment]
@kuujo PTAL

File: protocols/raft/src/main/java/io/atomix/protocols/raft/proxy/impl/DefaultRaftProxy.java [hunk 2]
```diff
@@ -68,6 +69,7 @@
   private volatile RaftProxyListener proxyListener;
   private volatile RaftProxyInvoker proxyInvoker;
   private volatile RaftProxyState state;
+  private final Consumer<NodeId> leaderChangeListener = this::onLeaderChange;
 
   public DefaultRaftProxy(
```
Comment (johnou) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/proxy/impl/DefaultRaftProxy.java]: This is called DiscreteRaftProxyClient in 2.0, correct?

**Comment by johnou** [comment]
Updated to include fixes from more-bug-fixes branch.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/roles/FollowerRole.java [hunk 3]
```diff
@@ -95,7 +95,7 @@ private void resetHeartbeatTimeout() {
     Duration delay = raft.getElectionTimeout().plus(Duration.ofMillis(random.nextInt((int) raft.getElectionTimeout().toMillis())));
     heartbeatTimer = raft.getThreadContext().schedule(delay, () -> {
       heartbeatTimer = null;
-      if (isRunning()) {
+      if (isRunning() && (raft.getFirstCommitIndex() == 0 || raft.getState() == RaftContext.State.READY)) {
```
Comment (johnou) [inline protocols/raft/src/main/java/io/atomix/protocols/raft/roles/FollowerRole.java]: @kuujo this needs a second look to make sure the merge is correct as failure detection was removed from FollowerRole in 2.1.

**Comment by johnou** [comment]
@kuujo didn't cherry-pick https://github.com/atomix/atomix/commit/c620efbeec568057c2055a6b6a9e70360afb31ee as PrimitiveException does not have ProtocolException and RaftException.ProtocolException belongs in atomix-raft which is not a dependency of atomix-primitve.

File: protocols/raft/src/main/java/io/atomix/protocols/raft/proxy/impl/DefaultRaftProxy.java (hunk 2 above)
Comment (kuujo) [reply protocols/raft/src/main/java/io/atomix/protocols/raft/proxy/impl/DefaultRaftProxy.java]: Yep

**Comment by kuujo** [comment]
I think this is good

**Pull request merged**